### PR TITLE
wip: Explore narwhals

### PR DESCRIPTION
**Pull request by MarcoGorelli**
the original branch I'd made went hopelessly out-of-date, so I've tried picking up from #189 again

Demo:
```python
In [1]: import pandas as pd
   ...: import polars as pl
   ...: 
   ...: from formulaic import model_matrix
   ...: 
   ...: df = pd.DataFrame({"a": [1,2,None], "b": list("abc"), "c": pd.Categorical(list("abc"))})
   ...: 
   ...: print(model_matrix("a + b + c", df, na_action="ignore"))
   ...: print(model_matrix("a + b + c", pl.from_pandas(df), na_action="ignore"))
   Intercept    a  b[T.b]  b[T.c]  c[T.b]  c[T.c]
0        1.0  1.0       0       0       0       0
1        1.0  2.0       1       0       1       0
2        1.0  NaN       0       1       0       1
shape: (3, 6)
┌───────────┬─────┬────────┬────────┬────────┬────────┐
│ Intercept ┆ a   ┆ b[T.b] ┆ b[T.c] ┆ c[T.b] ┆ c[T.c] │
│ ---       ┆ --- ┆ ---    ┆ ---    ┆ ---    ┆ ---    │
│ f64       ┆ f64 ┆ i64    ┆ i64    ┆ i64    ┆ i64    │
╞═══════════╪═════╪════════╪════════╪════════╪════════╡
│ 1.0       ┆ 1.0 ┆ 0      ┆ 0      ┆ 0      ┆ 0      │
│ 1.0       ┆ 2.0 ┆ 1      ┆ 0      ┆ 1      ┆ 0      │
│ 1.0       ┆ NaN ┆ 0      ┆ 1      ┆ 0      ┆ 1      │
└───────────┴─────┴────────┴────────┴────────┴────────┘
```

### CODE REVIEW

**Comment by stanmart** [comment]
This is awesome! 👏 Seeing how you implemented the `NarwhalsMaterializer` will be super helpful when we extend `TabamtMaterializer` for arbitrary dataframe inputs via narwhals. And having formulaic just work™ for polars dataframes feels like magic.

I noticed two minor issues when playing around with it:

- `NarwhalsMaterializer` does not seem to handle `na_action` correctly. For example, the following should drop a row, but does not:
  ```python
  In [1]: df = pd.DataFrame({
       ...:     "xm": np.concat([np.random.rand(7), [np.nan]]),
       ...: })
       ...: formulaic.model_matrix("xm", data=pl.from_pandas(df), na_action="drop")
  Out [1]:
  shape: (8, 2)
  ┌───────────┬──────────┐
  │ Intercept ┆ xm       │
  │ ---       ┆ ---      │
  │ f64       ┆ f64      │
  ╞═══════════╪══════════╡
  │ 1.0       ┆ 0.547805 │
  │ 1.0       ┆ 0.577707 │
  │ 1.0       ┆ 0.252147 │
  │ 1.0       ┆ 0.167533 │
  │ 1.0       ┆ 0.870474 │
  │ 1.0       ┆ 0.650435 │
  │ 1.0       ┆ 0.897262 │
  │ 1.0       ┆ NaN      │
  └───────────┴──────────┘
  ```
  `na_action="raise"` has similar issues, as well as using formats other than polars.

- You don't get back a narwhals dataframe unless you import narwhals as `narwhals.stable.v1`. It might be best practice to do so, but I would still expect the following to return an `nw.DataFrame`:
  ```python
  In [2]: formulaic.model_matrix("xm", data=narwhals.from_native(df))
  Out[2]:
     Intercept        xm
  0        1.0  0.547805
  1        1.0  0.577707
  2        1.0  0.252147
  3        1.0  0.167533
  4        1.0  0.870474
  5        1.0  0.650435
  6        1.0  0.897262
  7        1.0       NaN
  ```
  It works when I use the stable API, though:
  ```python
  In [3]: formulaic.model_matrix("xm", data=narwhals.stable.v1.from_native(df), na_action="drop")
  Out[3]:
  ┌───────────────────────────────────────┐
  | Narwhals DataFrame                    |
  | Use `.to_native` to see native output |
  └───────────────────────────────────────┘
  ```
  I think the culprit is that [this](https://github.com/MarcoGorelli/formulaic/blob/9464201db39016851d3747d7b96ea8de891fb11d/formulaic/materializers/narwhals.py#L205) isinstance sheck fails in the first case. (There are a few other `isinstance(..., nw.Series)` lines elsewhere too, but I have not look into if those can be problematic, too.

File: formulaic/materializers/narwhals.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+import narwhals.stable.v1 as nw
+import numpy
+import pandas
+import scipy.sparse as spsparse
+from interface_meta import override
+
+# Prefer these for `isinstance` checks so they catch both
+# narwhals.stable.v1.DataFrame and narwhals.DataFrame
+from narwhals import DataFrame as NwDataFrame
+from narwhals import Series as NwSeries
```
Comment (MarcoGorelli) [inline formulaic/materializers/narwhals.py:16]: I don't like this, we should address it in Narwhals - I've made an issue (https://github.com/narwhals-dev/narwhals/issues/1548), will update once that's addressed / released

Comment (MarcoGorelli) [reply formulaic/materializers/narwhals.py]: done

**Comment by MarcoGorelli** [comment]
thanks @stanmart for trying this out! have addressed

**Comment by matthewwardrop** [comment]
Thank you so much for pushing forward with these changes @MarcoGorelli ! I decided to push out 1.1.0 without a few new features, including this one, to unblock upstream usage in statsmodels. I'm still planning to fast follow with a 1.2.0 release over the Winter break, which should include these changes. I'll review these changes as soon as I can :).

**Comment by MarcoGorelli** [comment]
thanks!

do you plan to support Python3.7 much longer? it's not supported in Narwhals, so this is to help decide between working around py3.7 here or just waiting until it's dropped

**Comment by matthewwardrop** [comment]
Hi @MarcoGorelli ! I'm fine with dropping Python 3.7 and 3.8 in 1.2 (to track with the official EOL status of those versions). I'll push up the deprecation soon.

**Comment by matthewwardrop** [comment]
(This is now done in main).

**Comment by MarcoGorelli** [comment]
Sounds great, thanks 🙏! Please do feel free to reach out if anything stumps you

**Pull request merged**